File: src/main/java/me/chell/samsara/api/value/ValueBuilder.java
package me.chell.samsara.api.value;

import java.util.function.Predicate;

public class ValueBuilder<T> {
    private final String name;
    private final T value;
    private T sliderMin;
    private T sliderMax;
    private Predicate<Boolean> visible;

    public ValueBuilder(String name, T value) {
        this.name = name;
        this.value = value;
    }

    public ValueBuilder<T> bounds(T min, T max) {
        sliderMin = min;
        sliderMax = max;
        return this;
    }

    public ValueBuilder<T> visible(Predicate<Boolean> predicate) {
        visible = predicate;
        return this;
    }

    public Value<T> build() {
        if(visible == null) visible = b -> true;
        return new Value<>(name, value, sliderMin, sliderMax, visible);
    }
}

File: src/main/kotlin/me/chell/samsara/api/value/ValueBuilder.kt
package me.chell.samsara.api.value

import java.util.function.Predicate

class ValueBuilder<T>(private val name: String, private val value: T) {
    private var description: String = "No description."
    private var displayName = name
    private var min: T? = null
    private var max: T? = null
    private var predicate = Predicate<Boolean> {true}

    fun description(description: String): ValueBuilder<T> {
        this.description = description
        return this
    }

    fun bounds(min: T, max: T): ValueBuilder<T> {
        this.min = min
        this.max = max
        return this
    }

    fun name(name: String): ValueBuilder<T> {
        displayName = name
        return this
    }

    fun visible(predicate: Predicate<Boolean>): ValueBuilder<T> {
        this.predicate = predicate
        return this
    }

    fun build(): Value<T> = Value(name, value, description, displayName, min, max, predicate)
}